### PR TITLE
Upating schema to use standardized dataset features

### PR DESCRIPTION
**Pull request by aolofsson**
Standardizing on datasheet features.
Generic feature strings don't work...they make it impossible to write higher level portable software. 
Datasheeets and interface terms tend to be open, so I don't see a reason to not standardize on terms.

### CODE REVIEW

File: siliconcompiler/schema/schema_cfg.py [hunk 1]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+            sctype='str',
+            shorthelp="Datasheet: part number",
+            switch="-partnumber '<str>'",
+            example=[
+                "cli: -partnumber 'ZA001-P484'",
+                "api: chip.set('datasheet', 'partnumber', 'ZA001-PV84)"],
+            schelp="""A unique device identifier.""")
+
+    # Category
+    scparam(cfg, ['datasheet', 'category'],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:778]: would `type` be a more descriptive key?

File: siliconcompiler/schema/schema_cfg.py [hunk 2]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+            shorthelp="Datasheet: speed rating",
+            switch="-speed '<int>'",
+            example=[
+                "cli: -speed 100'",
+                "api: chip.set('datasheet', 'speed', 100')"],
+            schelp="""Device peak speed rating.""")
+
+    # Total SRAM
+    scparam(cfg, ['datasheet', 'sram'],
+            sctype='int',
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:942]: float

File: siliconcompiler/schema/schema_cfg.py [hunk 3]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+            switch="-iocount '<int>'",
+            example=[
+                "cli: -iocount 100'",
+                "api: chip.set('datasheet', 'iocount', 100)"],
+            schelp="""Device total number of I/Os (not counting supplies).""")
+
+    # Total OPS
+    scparam(cfg, ['datasheet', 'ops'],
+            sctype='float',
+            shorthelp="Datasheet: operations per second",
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:975]: units?

File: siliconcompiler/schema/schema_cfg.py [hunk 4]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+
+    metrics = {'speed': ['speed', 100, 'MHz'],
+               'width': ['width', 4, None],
+               'channels': ['channels', 4, None]
+               }
+
+    for i, v in metrics.items():
+        scparam(cfg, ['datasheet', 'io', name, i],
+                unit=v[2],
+                sctype='float',
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1024]: width and channels should be int not float

File: siliconcompiler/schema/schema_cfg.py [hunk 5]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+    scparam(cfg, ['datasheet', 'proc', name, 'datatypes'],
+            sctype='[enum]',
+            enum=['int1', 'int4', 'int8', 'int16', 'int32', 'int64', 'int128',
+                  'sint4', 'sint8', 'sint16', 'sint32', 'sint64', 'sint128',
+                  'bfloat16', 'fp16', 'fp32', 'fp64'],
+            shorthelp="Datasheet: processor datatypes",
+            switch="-proc_datatypes 'name <str>'",
+            example=[
+                "cli: -proc_datatypes '0 SIMD'",
+                "api: chip.set('datasheet','proc','cpu','features', 'SIMD')"],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1065]: example does not align with key

File: siliconcompiler/schema/schema_cfg.py [hunk 6]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+            schelp="""List of maker specified processor features.""")
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1066]: maker->manufacturer

File: siliconcompiler/schema/schema_cfg.py [hunk 7]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+
+    metrics = {'archsize': ['architecture size', 64, None],
+               'cores': ['number of cores', 4, None],
+               'speed': ['maximum frequency', 100, 'MHz'],
+               'ops': ['operations per cycle per core', 4, None],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1071]: are these not overlapping with other keys in the datasheet?

File: siliconcompiler/schema/schema_cfg.py [hunk 8]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+               'mults': ['hard multiplier units per core', 100, None],
+               'icache': ['l1 icache size', 32, 'KB'],
+               'dcache': ['l1 dcache size ', 32, 'KB'],
+               'l2cache': ['l2 cache size', 1024, 'KB'],
+               'l3cache': ['l3 cache size', 1024, 'KB'],
+               'sram': ['shared sram', 128, 'KB'],
+               'nvm': ['shared non-volatile memory', 128, 'KB']}
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1078]: same

File: siliconcompiler/schema/schema_cfg.py [hunk 9]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+                "cli: -fpga_arch 'openfpga'",
+                "api: chip.set('datasheet', 'fpga', 'arch', 'openfpga')"],
+            schelp="""FPGA architecture.
+            """)
+
+    metrics = {'luts': ['LUTs (4-input)', 32000, None],
+               'registers': ['registers', 100, None],
+               'plls': ['pll blocks', 1, None],
+               'mults': ['hard multiplier units', 100, None],
+               'totalram': ['totaldistributed ram', 128, 'Kb'],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1109]: totaldistributed -> total distributed

File: siliconcompiler/schema/schema_cfg.py [hunk 10]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+               'distram': ['distributed ram', 128, 'Kb'],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1110]: how is total distributed different from distram?

File: siliconcompiler/schema/schema_cfg.py [hunk 11]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+            sctype='[str]',
+            shorthelp="Datasheet: analog features",
+            switch="-analog_features 'name <str>'",
             example=[
-                "cli: -datasheet_feature 'ram 64e6'",
-                "api: chip.set('datasheet', 'feature', 'ram', 64e6)"],
-            schelp="""Quantity of a specified feature. The 'unit'
-            field should be used to specify the units used when unclear.""")
+                "cli: -analog_features '0 SIMD'",
+                "api: chip.set('datasheet','analog','adc0','features', 'differential input')"],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1143]: align examples.

File: siliconcompiler/schema/schema_cfg.py [hunk 12]
```diff
@@ -875,19 +1313,19 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
+                "cli: -pin_type 'vdd global supply'",
                 "api: chip.set('datasheet', 'pin', 'vdd', 'type', 'global', 'supply')"],
             schelp="""Pin type specified on a per mode basis.""")
 
     # Pin direction
     scparam(cfg, ['datasheet', 'pin', name, 'dir', mode],
             sctype='str',
             shorthelp="Datasheet: pin direction",
-            switch="-datasheet_pin_dir 'name mode <str>'",
+            switch="-pin_dir 'name mode <str>'",
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1326]: direction should be enum

**Review by gadfort — CHANGES_REQUESTED** [review]

File: siliconcompiler/schema/schema_cfg.py [hunk 13]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+            sctype='bool',
+            shorthelp="Datasheet: physical device indicator",
+            switch="-physical '<bool>'",
+            example=[
+                "cli: -physical True",
+                "api: chip.set('datasheet', 'physical', True)"],
+            schelp="""A physical (real) device. The alternative to a physical
+            device is a virtual device that has not been manufactured. Examples
+            of physical devices include dies, chips, and boards. Examples of
+            virtual devices include dsign RTL, soft code, hard macros, layout,
```
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:815]: fixed

File: siliconcompiler/schema/schema_cfg.py (hunk 1 above)
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:778]: agreed

File: siliconcompiler/schema/schema_cfg.py [hunk 14]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+            switch="-manufacturer '<str>'",
+            example=[
+                "cli: -manufacturer 'Zero ASIC'",
+                "api: chip.set('datasheet', 'manufacturer', 'Zero ASIC')"],
+            schelp="""Device manufacturer/vendor.""")
+
+    # Price
+    quantity = 'default'
+    scparam(cfg, ['datasheet', 'price'],
+            sctype='[(float,float)]',
```
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:842]: I would rather not introduce another testing corner case.

File: siliconcompiler/schema/schema_cfg.py [hunk 15]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+                  'last time buy', 'obsolete'],
+            shorthelp="Datasheet: product status",
+            switch="-status '<str>'",
+            example=[
+                "cli: -stauts 'active'",
+                "api: chip.set('datasheet', 'status', 'active')"],
+            schelp="""Device production status.""")
+
+    # Speed
+    scparam(cfg, ['datasheet', 'speed'],
```
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py]: changed to fmax

File: siliconcompiler/schema/schema_cfg.py (hunk 7 above)
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:1079]: somewhat. the 'proc' number is specified on a per block/core basis. For example, there can be a big and little core  runnning at different frequencies. The top level number is the "marketing" number for the device.

File: siliconcompiler/schema/schema_cfg.py (hunk 12 above)
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:1403]: Nice catch!

File: siliconcompiler/schema/schema_cfg.py (hunk 10 above)
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:1187]: fixed, total=dist+block

File: siliconcompiler/schema/schema_cfg.py (hunk 11 above)
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py]: fixed

File: siliconcompiler/schema/schema_cfg.py [hunk 16]
```diff
@@ -784,36 +784,34 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
                 "cli: -doc 'za001.pdf'",
                 "api: chip.set('datasheet', 'doc', 'za001.pdf)"],
             schelp="""Device datasheet document.""")
 
-    # Physical device (boolean)
-    scparam(cfg, ['datasheet', 'physical'],
-            sctype='bool',
-            shorthelp="Datasheet: physical device indicator",
-            switch="-physical '<bool>'",
+    # Device abstraction (boolean)
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:804]: not a boolean anymore

File: siliconcompiler/schema/schema_cfg.py [hunk 17]
```diff
@@ -969,18 +968,27 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
             schelp="""Device total number of I/Os (not counting supplies).""")
 
-    # Total OPS
-    scparam(cfg, ['datasheet', 'ops'],
-            sctype='float',
-            shorthelp="Datasheet: operations per second",
-            switch="-ops '<float>'",
+    # Total on-device RAM
+    scparam(cfg, ['datasheet', 'ram'],
+            sctype='int',
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:973]: float?

File: siliconcompiler/schema/schema_cfg.py [hunk 18]
```diff
@@ -1088,6 +1096,75 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
+                "cli: -memory_banks 'm0 4'",
+                "api: chip.set('datasheet', 'memory', 'm0', 'banks', 4)"],
+            schelp="""Memory banks.""")
+
+    # Timing
+    metrics = {'fmax': ['max frequency', (1e-9, 1e-9, 1e-9), 'Hz'],
+               'tcycle': ['access cycle time', (1e-9, 1e-9, 1e-9), 'Hz'],
+               'twr': ['write cycle time', (1e-9, 1e-9, 1e-9), 'ns'],
+               'trd': ['read cycle time', (1e-9, 1e-9, 1e-9), 'ns'],
+               'trefresh': ['refresh time', (100e-9, 100e-9, 100e-9), 'ns'],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1144]: 'ns' -> 's' or fix to 1.0 and 100.0

File: siliconcompiler/schema/schema_cfg.py [hunk 19]
```diff
@@ -1088,6 +1096,75 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
+            shorthelp="Datasheet: memory banks",
+            switch="-memory_banks 'name <int>'",
+            example=[
+                "cli: -memory_banks 'm0 4'",
+                "api: chip.set('datasheet', 'memory', 'm0', 'banks', 4)"],
+            schelp="""Memory banks.""")
+
+    # Timing
+    metrics = {'fmax': ['max frequency', (1e-9, 1e-9, 1e-9), 'Hz'],
+               'tcycle': ['access cycle time', (1e-9, 1e-9, 1e-9), 'Hz'],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1141]: tcycle -> 's'

File: siliconcompiler/schema/schema_cfg.py [hunk 20]
```diff
@@ -1088,6 +1096,75 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
+            sctype='int',
+            shorthelp="Datasheet: memory banks",
+            switch="-memory_banks 'name <int>'",
+            example=[
+                "cli: -memory_banks 'm0 4'",
+                "api: chip.set('datasheet', 'memory', 'm0', 'banks', 4)"],
+            schelp="""Memory banks.""")
+
+    # Timing
+    metrics = {'fmax': ['max frequency', (1e-9, 1e-9, 1e-9), 'Hz'],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1140]: 100e6 as example?

File: siliconcompiler/schema/schema_cfg.py [hunk 21]
```diff
@@ -1088,6 +1096,75 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
+               'tcycle': ['access cycle time', (1e-9, 1e-9, 1e-9), 'Hz'],
+               'twr': ['write cycle time', (1e-9, 1e-9, 1e-9), 'ns'],
+               'trd': ['read cycle time', (1e-9, 1e-9, 1e-9), 'ns'],
+               'trefresh': ['refresh time', (100e-9, 100e-9, 100e-9), 'ns'],
+               'terase': ['erase time', (100e-9, 100e-9, 100e-9), 's'],
+               'tcl': ['column address latency', (100e-9, 100e-9, 100e-9), 'cycles'],
+               'trcd': ['row address latency', (100e-9, 100e-9, 100e-9), 'cycles'],
+               'trp': ['row precharge time latency', (100e-9, 100e-9, 100e-9), 'cycles'],
+               'tras': ['row active time latency', (100e-9, 100e-9, 100e-9), 'cycles'],
+               'twearout': ['write/erase wear-out time', (1e-9, 1e-9, 1e-9), 'cycles'],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1150]: cycles are 'int'

File: siliconcompiler/schema/schema_cfg.py [hunk 22]
```diff
@@ -1088,6 +1096,75 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
+               'bwrd': ['maximum read bandwidth', (1e-9, 1e-9, 1e-9), 'bps'],
+               'bwwr': ['maximum write bandwidth', (1e-9, 1e-9, 1e-9), 'bps'],
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:1152]: fix example to be more realistic

File: siliconcompiler/schema/schema_cfg.py [hunk 23]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
-    # Device Features
-    scparam(cfg, ['datasheet', 'feature', name],
-            sctype='float',
+    # Part Number
+    scparam(cfg, ['datasheet', 'partnumber'],
+            sctype='str',
+            shorthelp="Datasheet: part number",
+            switch="-partnumber '<str>'",
+            example=[
+                "cli: -partnumber 'ZA001-P484'",
```
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py]: fixed

File: siliconcompiler/schema/schema_cfg.py [hunk 24]
```diff
@@ -1423,19 +1436,19 @@ def schema_datasheet(cfg, name='default', mode='default'):
...
+                "cli: -datasheet_pin_standard 'clk def LVCMOS'",
                 "api: chip.set('datasheet', 'pin', 'clk', 'standard', 'def', 'LVCMOS')"],
             schelp="""Pin electrical signaling standard (LVDS, LVCMOS, TTL, ...).""")
 
     # Pin interface map
     scparam(cfg, ['datasheet', 'pin', name, 'interface', mode],
             sctype='[str]',
             shorthelp="Datasheet: pin interface map",
-            switch="-pin_signal 'name mode <str>'",
+            switch="-datasheet_pin_signal 'name mode <str>'",
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py]: should this be `datasheet_pin_interface`?

File: siliconcompiler/schema/schema_cfg.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py]: removed from schema

File: siliconcompiler/schema/schema_cfg.py [hunk 25]
```diff
@@ -763,42 +764,465 @@ def schema_pdk(cfg, stackup='default'):
...
+            shorthelp="Datasheet: total device RAM",
+            switch="-sram '<int>'",
+            example=[
+                "cli: -sram 128'",
+                "api: chip.set('datasheet', 'sram', 128)"],
+            schelp="""Device total RAM.""")
+
+    # Total NVM
+    scparam(cfg, ['datasheet', 'nvm'],
+            sctype='int',
```
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py]: removed

**Comment by aolofsson** [comment]
hopefully we are good to go now.

**Review by gadfort — APPROVED** [review]
LGTM
